Source/NSLayoutConstraint.m

```diff
@@ -652,16 +654,89 @@ - (void) updateConstraints
 
 @implementation NSView (NSConstraintBasedLayoutInstallingConstraints)
 
+- (GSAutoLayoutEngine*) _getOrCreateLayoutEngine
+{
+  if (![self window])
+    {
+      return nil;
+    }
+  if (![[self window] _layoutEngine])
+    {
+      [[self window] _bootstrapAutoLayout];
+    }
+
+  return [[self window] _layoutEngine];
+ }
+
 - (void) addConstraint: (NSLayoutConstraint *)constraint
 {
-  // FIXME: Implement adding constraint to layout engine
+  if (![self _getOrCreateLayoutEngine])
+    {
+      return;
+    }
+
+  [[self _layoutEngine] addConstraint: constraint];
 }
 
 - (void) addConstraints: (NSArray*)constraints
 {
-  FOR_IN (NSLayoutConstraint*, constraint, constraints)
-    [self addConstraint: constraint];
-  END_FOR_IN (constraints);
+  if (![self _getOrCreateLayoutEngine])
+    {
+      return;
+    }
+
+  [[self _layoutEngine] addConstraints: constraints];
+}
+
+- (void) removeConstraint: (NSLayoutConstraint *)constraint
+{
+  if (![self _layoutEngine])
+    {
+      return;
+    }
+
+  [[self _layoutEngine] removeConstraint: constraint];
+}
+
+- (void) removeConstraints: (NSArray *)constraints
+{
+  if (![self _layoutEngine])
+    {
+      return;
+    }
+
+  [[self _layoutEngine] removeConstraints: constraints];
+}
+
+- (NSArray*) constraints
+{
+  GSAutoLayoutEngine *engine = [self _layoutEngine];
+  if (!engine)
+    {
+      return [NSArray array];
+    }
+
+  return [engine constraintsForView: self];
 }
 
 @end
+
+@implementation NSWindow (NSConstraintBasedLayoutCoreMethods)
+
+- (void) layoutIfNeeded
+{
+  [self updateConstraintsIfNeeded];
+  [[self contentView] _layoutViewAndSubViews];
+}
+
+- (void) updateConstraintsIfNeeded
+{
+  [[self contentView] updateConstraintsForSubtreeIfNeeded];
+}
+
+- (void) _bootstrapAutoLayout
+{
+  [self _setLayoutEngine: [[GSAutoLayoutEngine alloc] init]];
```

Comment:
Most likely you should AUTORELEASE this here.

Reply:
I'm not clear on the reason for this. The layout engine should exist for the lifetime of the NSWindow. From my understanding (or lack of) the alloc init would have a +1 for the engine, which would be decremented to zero when the NSWindow is deallocated. Please explain.

Reply:
Your thinking is correct here. The code as you wrote it is not a memory leak. The issue here is that you don't follow the established reference handling conventions of Objective-C. This could be confusing for readers of the code. One way to avoid this is to add some comments here and in the NSWindow code. Another way would be to stay with the expectations.

Reply:
No worries, fixed, thank you for the fast feedback while I'm still getting to grips with the conventions.